### ISSUE TITLE
win32eventreactor error on reactor.stop()

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11407]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| justinj reported|
|-|-|
|Trac ID|trac#379|
|Type|defect|
|Created|2003-10-31 22:22:03Z|

Attachments:

* [winsvc.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c82/c829eb96cef056a9003d7ab56ed6072e99089985/177b43fc4881925aecefae74fb5907bcc54ff1be.py) (991 bytes) - added by justinj on 2003-10-31 22:22:02Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__379 379
type__defect defect
reporter__justinj justinj
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__win32 win32
time__1067638923000000 1067638923000000
changetime__1469139705814797 1469139705814797
version__ 
owner__ 
cc__spiv cc__justinj
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
I'm submitting this as a bug per Itamar.

I setup a NT service, importing the win32eventreactor and
installing it.  The service starts up fine, but when I stop the service I
get the following in the logs.  The "15:08 ... Unexpected..." entry only
gets dumped in the log when stopping the service.

I've attached winsvc.py.

2003/10/30 15:07 Eastern Standard Time [-] Log opened.
2003/10/30 15:07 Eastern Standard Time [-]
twisted.spread.pb.PBServerFactory starting on 8787
2003/10/30 15:07 Eastern Standard Time [-] Starting factory
&lt;twisted.spread.pb.PBServerFactory instance at 0x00DF8D28>
2003/10/30 15:08 Eastern Standard Time [-] Unexpected error in main loop.
2003/10/30 15:08 Eastern Standard Time [-] Traceback (most recent call
last):
          File "E:\Python22\Lib\site-packages\win32\lib\win32serviceutil.py",
line 635, in SvcRun
            self.SvcDoRun()
          File "E:\ccase\python\winsvc.py", line 24, in SvcDoRun
            reactor.run()
          File "E:\Python22\Lib\site-packages\twisted\internet\default.py", line
122, in run
            self.mainLoop()
        --- &lt;exception caught here> ---
          File "E:\Python22\Lib\site-packages\twisted\internet\default.py", line
133, in mainLoop
            self.doIteration(t)
          File
"E:\Python22\Lib\site-packages\twisted\internet\win32eventreactor.py", line 180,
in doWaitForMultipleEvents
            fd, action = events[handles[val - WAIT_OBJECT_0]]
        exceptions.KeyError: &lt;PyHANDLE:224>
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Unfortunately, I don't have a windows machine to test this on (although your
test-case isn't complete anyway...).

I'd try changing SvcStop to call
    reactor.callFromThread(reactor.stop)
instead of just "reactor.stop()".  Can you try that and let us know if it fixes
it?  Thanks!
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
This is fixed now, without making any code changes.  I'm running 1.1.1 on
Windows 2000.  I don't recall when it was fixed, but it was some time ago.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Even better :)

Marking as resolved.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|GitHub commented|
|-|-|

In changeset 7d194cd0487d625e2806f9d307a0c1bc5d8201bf
```
#!CommitTicketReference repository="" revision="7d194cd0487d625e2806f9d307a0c1bc5d8201bf"
Merge pull request #379 from twisted/8460-rodrigc-words-notypes

Author: rodrigc
Reviewer: markrwilliams
Fixes: #8460

Remove use of types module in twisted/words
```
